### PR TITLE
refactor fork choice

### PR DESCRIPTION
**Pull request by bw-solana**
#### Problem
This code is complex by necessity, but also really hard to parse due to long functions and comments.

#### Summary of Changes

1. Break up `select_vote_and_reset_forks` into smaller sub-pieces
2. Un-nest a lot of things using let/else {early return} pattern
3. re-shuffle some things
4. flip polarity on some things
5. make some variable names more precise
6. add additional comments

No functional change is intended to be made

### CODE REVIEW

File: core/src/consensus/fork_choice.rs [hunk 1]
```diff
@@ -22,6 +22,16 @@ pub struct SelectVoteAndResetForkResult {
     pub heaviest_fork_failures: Vec<HeaviestForkFailures>,
 }
 
+struct CandidateVoteAndResetBanks<'a> {
```
Comment (bw-solana) [inline core/src/consensus/fork_choice.rs:25]: Pulled this struct definition outside of the `select_vote_and_reset_forks` function

File: core/src/consensus/fork_choice.rs [hunk 2]
```diff
@@ -54,6 +64,35 @@ pub trait ForkChoice {
     ) -> Vec<Self::ForkChoiceKey>;
 }
 
+fn last_vote_able_to_land(
```
Comment (bw-solana) [inline core/src/consensus/fork_choice.rs:67]: Pulled this out of the select_forks_failed_switch_threshold function, flipped the polarity, un-nested things

File: core/src/consensus/fork_choice.rs [hunk 3]
```diff
@@ -54,6 +64,35 @@ pub trait ForkChoice {
...
+    reset_bank: Option<&Bank>,
+    progress: &ProgressMap,
+    tower: &Tower,
+) -> bool {
+    let Some(heaviest_bank_on_same_voted_fork) = reset_bank else {
```
Comment (bw-solana) [inline core/src/consensus/fork_choice.rs:72]: It would be nice to add comments to each return conditional. I don't fully understand how to describe the reasoning concisely tho

File: core/src/consensus/fork_choice.rs [hunk 4]
```diff
@@ -64,31 +103,7 @@ fn select_forks_failed_switch_threshold(
...
-                    }
-                }
-                None => false,
-            }
-        }
-        None => false,
-    };
-
-    if last_vote_unable_to_land {
+    if !last_vote_able_to_land(reset_bank, progress, tower) {
```
Comment (bw-solana) [inline core/src/consensus/fork_choice.rs:113]: collapsed this down and early return (instead of if/else) to reduce indent for the more complex else leg

File: core/src/consensus/fork_choice.rs (hunk 3 above)
Comment (AshwinSekar) [reply core/src/consensus/fork_choice.rs:72]: The idea here is that if we are waiting to switch due to `FailedSwitchThreshold`, `reset_bank` will be on the same fork as our `last_voted_slot`. We can then use the bank to figure out if our last vote has landed.

This can be `None` if we have not voted before, or the fork from `last_voted_slot` is undergoing dump and repair due to duplicate consensus.
For these cases we don't expect to be stuck long enough for super refresh to be needed so no reason to check the conditions.

Comment (AshwinSekar) [reply core/src/consensus/fork_choice.rs:72]: maybe something like:
```rust
// No reset slot - we are in the middle of dump & repair, last vote landing is irrelevant
```

File: core/src/consensus/fork_choice.rs [hunk 5]
```diff
@@ -54,6 +64,35 @@ pub trait ForkChoice {
...
+        return true;
+    };
+
+    let Some(my_latest_landed_vote_slot) =
+        progress.my_latest_landed_vote(heaviest_bank_on_same_voted_fork.slot())
+    else {
+        return true;
+    };
+
+    // Last vote did not land. Now let's check if it's able to land.
```
Comment (AshwinSekar) [inline core/src/consensus/fork_choice.rs]: Suggestion for the comments
```
// Check if our last vote is able to land in order to determine if we should super refresh to vote at the tip
// Last vote has landed
    my_latest_landed_vote_slot >= last_voted_slot
        // Already voting at the tip
                || last_voted_slot >= heaviest_bank_on_same_voted_fork.slot()
        // Last vote is withink slot hashes, regular refresh is enough
                || heaviest_bank_on_same_voted_fork
            .is_in_slot_hashes_history(&last_voted_slot)
```

File: core/src/consensus/fork_choice.rs [hunk 6]
```diff
@@ -64,31 +103,7 @@ fn select_forks_failed_switch_threshold(
...
-            }
-        }
-        None => false,
-    };
-
-    if last_vote_unable_to_land {
+    if !last_vote_able_to_land(reset_bank, progress, tower) {
         // If we reach here, these assumptions are true:
         // 1. We can't switch because of threshold
         // 2. Our last vote was on a non-duplicate/confirmed slot
```
Comment (AshwinSekar) [inline core/src/consensus/fork_choice.rs]: nit: this is actually not true, this still works the same if our last vote was for a duplicate slot.
I can also address this in another PR I know it's not part of the refactor.

File: core/src/consensus/fork_choice.rs [hunk 7]
```diff
@@ -99,47 +114,196 @@ fn select_forks_failed_switch_threshold(
...
         // slots, it might be enough to get majority vote.
-        SwitchForkDecision::SameFork
-    } else {
-        // If we can't switch and our last vote was on a non-duplicate/confirmed slot, then
-        // reset to the the next votable bank on the same fork as our last vote,
-        // but don't vote.
+        return SwitchForkDecision::SameFork;
+    }
+
+    // If we can't switch and our last vote was on a non-duplicate/confirmed slot, then
```
Comment (AshwinSekar) [inline core/src/consensus/fork_choice.rs]: Same comment here about the non-duplicate/confirmed slot, can still be for a duplicate slot.

File: core/src/consensus/fork_choice.rs [hunk 8]
```diff
@@ -99,47 +114,196 @@ fn select_forks_failed_switch_threshold(
...
+                heaviest_bank_on_same_voted_fork.map(|bank| bank.as_ref()),
+                progress,
+                tower,
+                heaviest_bank.slot(),
+                failure_reasons,
+                switch_proof_stake,
+                total_stake,
+                initial_switch_fork_decision,
+            );
+            let candidate_vote_bank = if final_switch_fork_decision.can_vote() {
```
Comment (AshwinSekar) [inline core/src/consensus/fork_choice.rs]: I think we dropped some comments that were here previously

File: core/src/consensus/fork_choice.rs [hunk 9]
```diff
@@ -99,47 +114,196 @@ fn select_forks_failed_switch_threshold(
...
+    heaviest_bank: &'a Arc<Bank>,
+    heaviest_bank_on_same_voted_fork: Option<&'a Arc<Bank>>,
+    progress: &'a ProgressMap,
+    tower: &'a Tower,
+    failure_reasons: &mut Vec<HeaviestForkFailures>,
+    initial_switch_fork_decision: SwitchForkDecision,
+) -> CandidateVoteAndResetBanks<'a> {
+    match initial_switch_fork_decision {
+        SwitchForkDecision::FailedSwitchThreshold(switch_proof_stake, total_stake) => {
+            // Even though we think we want to switch to a new, heavier fork, we
```
Comment (AshwinSekar) [inline core/src/consensus/fork_choice.rs]: Would replace with:
```
If our last vote is unable to land (even through normal refresh), then we temporarily "super" refresh our vote to the tip of our last voted fork.
```

File: core/src/consensus/fork_choice.rs [hunk 10]
```diff
@@ -99,47 +114,196 @@ fn select_forks_failed_switch_threshold(
...
+            } else {
+                Some(heaviest_bank)
+            };
+            CandidateVoteAndResetBanks {
+                candidate_vote_bank,
+                reset_bank: heaviest_bank_on_same_voted_fork,
+                switch_fork_decision: final_switch_fork_decision,
+            }
+        }
+        SwitchForkDecision::FailedSwitchDuplicateRollback(latest_duplicate_ancestor) => {
```
Comment (AshwinSekar) [inline core/src/consensus/fork_choice.rs:297]: Also dropped some comments here

File: core/src/consensus/fork_choice.rs [hunk 11]
```diff
@@ -99,47 +114,196 @@ fn select_forks_failed_switch_threshold(
...
+                0, // In this case we never actually performed the switch check, 0 for now
+                0,
+            ));
+            CandidateVoteAndResetBanks {
+                candidate_vote_bank: None,
+                reset_bank,
+                switch_fork_decision: initial_switch_fork_decision,
+            }
+        }
+        _ => CandidateVoteAndResetBanks {
```
Comment (AshwinSekar) [inline core/src/consensus/fork_choice.rs]: I know you just copy pasted this but I think we should use:
```rust
SwitchForkDecision::SameFork | SwitchForkDecision::SwitchProof(_) => ...
```
and avoid the wildcard

File: core/src/consensus/fork_choice.rs [hunk 12]
```diff
@@ -179,221 +343,60 @@ pub fn select_vote_and_reset_forks(
...
-                propagated_stats.is_leader_slot,
-                fork_stats.fork_weight(),
-                fork_stats.total_stake,
-                propagated_stats.total_epoch_stake,
-            )
+    let Some(candidate_vote_bank) = candidate_vote_bank else {
+        // No viable candidate to vote on.
+        return SelectVoteAndResetForkResult {
+            vote_bank: None,
+            reset_bank: reset_bank.cloned(),
```
Comment (AshwinSekar) [inline core/src/consensus/fork_choice.rs:470]: Just noting that you combined the two else cases here, but I believe this is correct, no need to case on whether `reset_bank` is `Some`

**Review by AshwinSekar — COMMENTED** [review]
structure looks good just some small comments

File: core/src/consensus/fork_choice.rs (hunk 6 above)
Comment (bw-solana) [reply core/src/consensus/fork_choice.rs]: I'll remove assumption 2 from the comment

File: core/src/consensus/fork_choice.rs (hunk 7 above)
Comment (bw-solana) [reply core/src/consensus/fork_choice.rs]: trimmed comment to:
```
// If we can't switch, then reset to the the next votable bank on the same
// fork as our last vote, but don't vote.
```

File: core/src/consensus/fork_choice.rs (hunk 8 above)
Comment (bw-solana) [reply core/src/consensus/fork_choice.rs]: Whoops. Updated the old comments to the following:
```
// We need to "super" refresh our vote to the tip of our last
// voted fork because our last vote is unable to land. This is
// inferred by initially determining we can't vote but then
// determining we can vote on the same fork.
```
and:
```
// Just return the original vote candidate (the heaviest bank)
// for logging purposes. We can't actually vote on it, but this
// will allow us to check if there are any additional voting
// failures besides the switch threshold.
```

File: core/src/consensus/fork_choice.rs (hunk 10 above)
Comment (bw-solana) [reply core/src/consensus/fork_choice.rs:297]: Added back in and tweaked the wording slightly.

This also triggered me to create sub-functions for each leg of the match because the comments are so giant that the `select_candidate_vote_and_reset_banks` function becomes huge.

File: core/src/consensus/fork_choice.rs (hunk 12 above)
Comment (bw-solana) [reply core/src/consensus/fork_choice.rs:470]: Good eye. Yeah, I was thinking we didn't need two different else legs. We can just clone the None

File: core/src/consensus/fork_choice.rs [hunk 13]
```diff
@@ -54,7 +64,47 @@ pub trait ForkChoice {
...
+    };
+
+    let Some(last_voted_slot) = tower.last_voted_slot() else {
+        // No previous vote.
+        return true;
+    };
+
+    let Some(my_latest_landed_vote_slot) =
+        progress.my_latest_landed_vote(heaviest_bank_on_same_voted_fork.slot())
+    else {
```
Comment (bw-solana) [inline core/src/consensus/fork_choice.rs:85]: @AshwinSekar What would hitting this else case mean? We've never landed a vote or we've purged that fork?

File: core/src/consensus/fork_choice.rs [hunk 14]
```diff
@@ -54,7 +64,47 @@ pub trait ForkChoice {
     ) -> Vec<Self::ForkChoiceKey>;
 }
 
-fn select_forks_failed_switch_threshold(
+/*The idea here is that if we are waiting to switch due to FailedSwitchThreshold, reset_bank will be on the same fork as our last_voted_slot.
```
Comment (AshwinSekar) [inline core/src/consensus/fork_choice.rs]: accidentally left this in?

File: core/src/consensus/fork_choice.rs (hunk 13 above)
Comment (AshwinSekar) [reply core/src/consensus/fork_choice.rs:85]: exactly we've either:
- never landed a vote (non voting validator) no reason to super refresh.
- fork has been pruned or is in the middle of dump & repair, so no need to super refresh.

**Review by AshwinSekar — COMMENTED** [review]
almost there, just 2 comments

File: core/src/consensus/fork_choice.rs (hunk 14 above)
Comment (bw-solana) [reply core/src/consensus/fork_choice.rs]: yeah, whoops

**Review by AshwinSekar — APPROVED** [review]
thanks for doing this! LGTM
